README

                            bacKardi 

    *   Aplicación para realizar Backup en sistemas tipo Unix   *


                    Daniel Martin Maldonado     <danyx@elcodigok.com.ar>
                                                <danyx@larepaweb.com.ar>

Descripcion General
--------------------------------------------------------------------------------

  Este programa es una implementación simple y rápida para la realizar de Backup
  en sistemas tipo Unix.

  A través del mismo es posible realizar tanto Backup completo, incrementales
  como asi tambien restaurar los mismos, ver los archivos que contienen, etc.


Requerimientos
--------------------------------------------------------------------------------

  * GNU/Linux
        Cualquier distribucion instalada.

  * tar

  * gzip

  * bzip2

  * unrar

  * unzip

  * arj


Uso
--------------------------------------------------------------------------------

  Para poder ejecutar backardi es necesario que seas super usuario root y se va
  a encontrar el script en /sbin/backardi.

  El directorio de configuracion general se va a encontrar en /etc/backardi en el 
  vas a encotrar todas las configuraciones referentes al funcionamiento del sistema 
  de backup.



NEWS
0.7

    - Organizacion y estandarizacion de los archivos.
    - Se incorpora el archivo VERSION
    - Es posible configurar el formato de la fecha para el nombre de los 
      backups.

0.6

    - Optimizacion y estandarizacion en el código fuente de bacKardi

0.5

    - Implementacion del unistall para eliminar backardi de la PC.

0.4

    - Implementacion de paginas de ayuda y manual de usuario.

0.3

    - Implementacion de un instalador de la aplicacion.

0.2

    - Implementacion de algoritmos de compresion rar, tar, tar.gz, tar.bz2 y arj.

0.1

    - Primer trabajo para la version beta de bacKardi. En primera instancia solo en lenguaje Espanol. Armado de la estructura jerarquica de directorios.


install.sh
#!/bin/bash
# ------------------------------------------------------------
# Sistema de Backup: bacKardi
#
# Autor:    Maldonado Daniel Martin
#
#			<danyx@elcodigok.com.ar>
#			<danyx@larepaweb.com.ar>
#
# fecha:    06/07/2010
#
# Instalador del sistema
#
# ------------------------------------------------------------ 

# ------------------- Funciones de Sistema -------------------

which_cmd() {
  local cmd=`which $2 2>/dev/null | head -n 1`
  eval "${1}"="${cmd}"
  return 0
}

which_compress() {
  local cmd=`which $1 2>/dev/null | head -n 1`
  if [ $? -gt 0 -o ! -x "${cmd}" ] ; then {
	"${ECHO}" -e "[\033[01;31m X \033[00m]── "${2}""
    return 1
  } else {
    "${ECHO}" -e "[\033[01;32m O \033[00m]── "${2}"	"${cmd}""
    return 0
  }
  fi
}

mensaje() {
  local mensaje="${1}"
  "${ECHO}" -e "${mensaje}"
}

exito() {
  local mensaje="${1}"
  "${ECHO}" -e "${mensaje}\t" "\033[01;31m [ OK ] \033[00m"
}

fracaso() {
  local mensaje="${1}"
  "${ECHO}" -e "${mensaje}\t" "\033[01;31m [ FRACASO ] \033[00m"
}

titulo() {
  local mensaje="${1}"
  "${ECHO}" -e "[ "${mensaje}" ]"
}

mensaje_exito() {
  local mensaje="${1}"
  "${ECHO}" -e "[\033[01;32m OK \033[00m] ..  " "${mensaje}"
}

mensaje_fracaso() {
  local mensaje="${1}"
  "${ECHO}" -e "[\033[01;34m ! \033[00m] ... " "${mensaje}"
}

# ------------------- Variables de sistema -------------------

which_cmd   ECHO    echo
which_cmd   CP      cp
which_cmd   ID      id
which_cmd   MKDIR   mkdir
which_cmd   LN      ln
which_cmd   PWD     pwd

DIR_DESTINO="/etc/backardi/"
DIR_ENLACE="/usr/sbin"
DIR_MAN="/usr/share/man/man1"
DIR_TRABAJO=`"${PWD}"`
SISTEMA="bacKardi"
VERSION="0.7"

# --------------- Control de usuario ------------------------

# La ejecucion del script solo la podra realizar el super 
# usuario root
if (( `"${ID}" -u` != 0)); then {
  mensaje "No puede ser iniciado "${SISTEMA}" v"${VERSION}": Permiso denegado (usted debe ser root)."
  fracaso "Verifique el usuario y luego intente nuevamente."
	exit;
} fi

titulo  "INSTALACION DEL SISTEMA "${SISTEMA}" v"${VERSION}""

# ----- Crear directorio de trabajo en el sistema------------

if ( [ -e "${DIR_DESTINO}" ] ) ; then {
  mensaje_fracaso "No se puede crear el directorio "${DIR_DESTINO}" por que ya existe."
} else {
	"${MKDIR}" "${DIR_DESTINO}"
  if [ $? -eq 0 ]; then {
    mensaje_exito "Directorio "${DIR_DESTINO}.""
  }
  fi
}
fi

# ----------- Copiar archivo ejecutable ---------------------

if ( [ -e "${DIR_DESTINO}"/backardi.sh ] ) ; then {
  mensaje_fracaso "No se puede copiar el archivo en "${DIR_DESTINO}"backardi.sh por que ya existe."
} else {
	#"${CP}" backardi/backardi.sh "${DIR_DESTINO}"
  "${CP}" "${DIR_TRABAJO}"/bin/backardi.sh "${DIR_DESTINO}"
  if [ $? -eq 0 ]; then {
    mensaje_exito "Copiado "${DIR_DESTINO}"backardi.sh"
  }
  fi
}
fi

# ------- Copiar archivo de configuracion -------------------

if ( [ -e "${DIR_DESTINO}"/bki.cnf ] ) ; then {
  mensaje_fracaso "No se puede copiar el archivo en "${DIR_DESTINO}"bki.cnf por que ya existe."
} else {
	#"${CP}" backardi/bki.cnf "${DIR_DESTINO}"
  "${CP}" "${DIR_TRABAJO}"/conf/bki.cnf "${DIR_DESTINO}"
  if [ $? -eq 0 ]; then {
    mensaje_exito "Copiado "${DIR_DESTINO}"bki.cnf"
  }
  fi
}
fi

# ------------- Crear enlace simbolico ---------------------

if ( [ -e "${DIR_ENLACE}"/backardi ] ) ; then {
  mensaje_fracaso "No se puede crear el enlace simbolico en "${DIR_ENLACE}"/backardi por que ya existe."
} else {
	#$LN --symbolic $DIR_DESTINO/backardi.sh $DIR_ENLACE/backardi
  "${LN}" --symbolic "$DIR_DESTINO"/backardi.sh "${DIR_ENLACE}"/backardi
  if [ $? -eq 0 ]; then {
    mensaje_exito "Copiado "${DIR_ENLACE}"/backardi"
  }
  fi
}
fi

# ----------- Agregar paginas del manual ------------------

if ( [ -e "${DIR_MAN}"/backardi.1.gz ] ) ; then {
  mensaje_fracaso "No se puede copiar el manual en "${DIR_MAN}"/backardi.1.gz por que ya existe."
} else {
	#"${CP}" backardi/backardi.1.gz "${DIR_MAN}"
  "${CP}" "${DIR_TRABAJO}"/man/backardi.1.gz "${DIR_MAN}"
  if [ $? -eq 0 ]; then {
    mensaje_exito "Copiado "${DIR_MAN}"/backardi.1.gz"
  }
  fi
}
fi

# ---------- Comprobar algoritmos de compresion -----------

titulo "COMPROBANDO SISTEMAS DE COMPRESIÓN"

which_compress      tar     "TAR"
which_compress      gzip    "GNUZip"
which_compress      bzip2   "Bzip2"
which_compress      rar     "RAR"
which_compress      arj     "ARJ"


uninstall.sh
#!/bin/bash
# ------------------------------------------------------------
# Sistema de Backup: bacKardi
#
# Autor:    Maldonado Daniel Martin
#
#       <danyx@elcodigok.com.ar>
#		    <danyx@larepaweb.com.ar>
#
# fecha:    07/07/2010
#
# ------------------------------------------------------------

# ----------------- Funciones de Sistema ---------------------

which_cmd() {
  local cmd=`which $2 2>/dev/null | head -n 1`
  eval "${1}"="${cmd}"
  return 0
}

mensaje() {
  local mensaje="${1}"
  "${ECHO}" -e "${mensaje}"
}

exito() {
  local mensaje="${1}"
  "${ECHO}" -e "${mensaje}\t" "\033[01;31m [ OK ] \033[00m"
}

fracaso() {
  local mensaje="${1}"
  "${ECHO}" -e "${mensaje}\t" "\033[01;31m [ FRACASO ] \033[00m"
}

titulo() {
  local mensaje="${1}"
  "${ECHO}" -e "[ "${mensaje}" ]"
}

mensaje_exito() {
  local mensaje="${1}"
  "${ECHO}" -e "[\033[01;32m OK \033[00m] ..  " "${mensaje}"
}

mensaje_fracaso() {
  local mensaje="${1}"
  "${ECHO}" -e "[\033[01;34m ! \033[00m] ... " "${mensaje}"
}

# ----------------- Variables del sistema --------------------

which_cmd   ECHO    echo
which_cmd   CP      cp
which_cmd   RM      rm
which_cmd   ID      id

SISTEMA="bacKardi"
VERSION="0.7"
DIR_INSTALL="/etc/backardi"
DIR_MAN_INSTALL="/usr/share/man/man1"
DIR_LN_INSTALL="/usr/sbin"

# --------------- Control de usuario ------------------------

# La ejecucion del script solo la podra realizar el super 
# usuario root
if (( `"${ID}" -u` != 0)); then {
	mensaje "No puede ser iniciado "${SISTEMA}" v"${VERSION}": Permiso denegado (usted debe ser root)."
	fracaso "Verifique el usuario y luego intente nuevamente.";
	exit;
} fi

titulo "ELIMINACIÓN DEL SISTEMA "${SISTEMA}" v"${VERSION}""

# ------------- Eliminar enlace simbolico ---------------------

if ( [ -e $DIR_LN_INSTALL/backardi ] ) ; then {
	$RM $DIR_LN_INSTALL/backardi
  if [ $? -eq 0 ]; then {
    mensaje_exito   "Se eliminó "${DIR_LN_INSTALL}"/backardi"
  }
  fi
} else {
  mensaje_fracaso   "No se puede eliminar el enlace "${DIR_LN_INSTALL}"/backardi"
}
fi

# ----- Eliminar directorio de trabajo en el sistema------------

if ( [ -e $DIR_INSTALL ] ) ; then {
	$RM -r $DIR_INSTALL
  if [ $? -eq 0 ]; then {
    mensaje_exito   "Se eliminó "${DIR_INSTALL}""
    mensaje_exito   "Se eliminó "${DIR_INSTALL}"backardi.sh"
    mensaje_exito   "Se eliminó "${DIR_INSTALL}"bck.cnf"
  }
  fi
} else {
  mensaje_fracaso   "No se puede eliminar el directorio "${DIR_INSTALL}""
}
fi

# ----------- Eliminar paginas del manual ------------------

if ( [ -e $DIR_MAN_INSTALL/backardi.1.gz ] ) ; then {
	$RM $DIR_MAN_INSTALL/backardi.1.gz
  if [ $? -eq 0 ]; then {
    mensaje_exito   "Se eliminó "${DIR_MAN_INSTALL}"/backardi.1.gz"
  }
  fi
} else {
  mensaje_fracaso   "No se puede eliminar el Manual "${DIR_MAN_INSTALL}"/backardi.1.gz"
}
fi


backardi/backardi.sh
#!/bin/bash 
# ------------------------------------------------------------
# Sistema de Backup: bacKardi
#
# Autor:    Maldonado Daniel Martin
#
#     <danyx@elcodigok.com.ar>
#     <danyx@larepaweb.com.ar>
#
# fecha:    08/07/2010
#
# Con este sistema es posible realizar backup tanto completos
# como asi tambien incrementales tanto de archivos como asi
# tambien de directorios.
# ------------------------------------------------------------

# ---------------- Funciones de Sistema ----------------------

which_cmd() {
  local cmd=`which $2 2>/dev/null | head -n 1`
  eval "${1}"="${cmd}"
  return 0
}

which_config() {
  local variable=`"${GREP}" -v "#" "${DIR_TRABAJO}"/bki.cnf | "${GREP}" "${2}" | "${SED}" 's/.*=//;s/^ *//' | head -n 1`
  eval "${1}"="${variable}"
  return 0
}

mensaje() {
  local mensaje="${1}"
  "${ECHO}" -e "${mensaje}"
  return 0
}

exito() {
  local mensaje="${1}"
  "${ECHO}" -e "${mensaje}\t" "\033[01;31m [ OK ] \033[00m"
  return 0
}

fracaso() {
  local mensaje="${1}"
  "${ECHO}" -e "${mensaje}\t" "\033[01;31m [ FRACASO ] \033[00m"
  return 1
}

# ----------------- Variables de Sistema ---------------------

which_cmd   GREP        grep
which_cmd   SED         sed
which_cmd   ID          id
which_cmd   ECHO        echo
which_cmd   HOSTNAME    hostname

# ---------- Obtener valores de configuracion ----------------

DIR_TRABAJO="/etc/backardi"

which_config    SISTEMA     sistema
which_config    VERSION     version
which_config    BACKUP_DIR  backup_dir
which_config    BACKUP_DEST_DIR     backup_dest_dir
which_config    COMPRESION  compresion
which_config    FORMATO     formato

#NOMBRE_FECHA=`date '+%A-%d-%B-%Y'`
NOMBRE_FECHA=`date +"${FORMATO}"`

#echo "Formato de la fecha: " `date +"${FORMATO}"`

# --------------- Control de usuario ------------------------

# La ejecucion del script solo la podra realizar el super 
# usuario root
if (( `"${ID}" -u` != 0)); then {
  mensaje   "No puede ser iniciado "${SISTEMA}" v"${VERSION}": Permiso denegado (usted debe ser root)."
  fracaso   "Verifique el usuario y luego intente nuevamente."
	exit;
} fi

# -------------- Control de aplicaciones ------------------

controlCompresor() {
	if ! ( [ -e /bin/$COMPRESION ] || [ -e /usr/bin/$COMPRESION ] ) ; then {
      mensaje   "El sistema de Compresion de archivos "${COMPRESION}" NO se encuentra disponible en su sistema."
      fracaso   "Puede utilizar otros sistemas de Compresion."
		exit;
	} fi
}

# ----- Funcion para determinar las extensiones -----------

controlExtension(){
  case "${COMPRESION}" in
    "tar")
      EXTENSION="tar"
		  COMANDO_COMPRIMIR="tar cvf"
		  COMANDO_DESCOMPRIMIR="tar xvf"
		  COMANDO_CONTENIDO="tar tvf"
		  ARGUMENTO_INCREMENTAL="-T -"
		  ARGUMENTO_DESCOMPRIMIR="-C"
		  TODOS=""
		  BANDERA=0
    ;;
    "gzip")
		  EXTENSION="tar.gz"
		  COMANDO_COMPRIMIR="tar vczf"
		  COMANDO_DESCOMPRIMIR="tar zxvf"
		  COMANDO_CONTENIDO="tar tzvf"
		  ARGUMENTO_INCREMENTAL="-T -"
		  ARGUMENTO_DESCOMPRIMIR="-C"
		  TODOS=""
		  BANDERA=0
    ;;
    "bzip2")
  		EXTENSION="tar.bz2"
	  	COMANDO_COMPRIMIR="tar -c"
		  COMANDO_DESCOMPRIMIR="tar zxvf"
      COMANDO_CONTENIDO=""
      ARGUMENTO_INCREMENTAL=""
		  ARGUMENTO_DESCOMPRIMIR="-C"
		  TODOS=""
 		  BANDERA=2
    ;;
    "rar")
  		EXTENSION="rar"
	  	COMANDO_COMPRIMIR="rar a"
		  COMANDO_DESCOMPRIMIR="rar x"
		  COMANDO_CONTENIDO="rar l"
		  ARGUMENTO_INCREMENTAL="/*"
		  ARGUMENTO_DESCOMPRIMIR="-C"
		  TODOS=""
		  BANDERA=1
		  $ECHO $COMANDO_COMPRIMIR $BACKUP_DEST_DIR/`$HOSTNAME`-$NOMBRE_FECHA-INC.$EXTENSION `find $BACKUP_DIR/* -mtime -1 -type f -print`
    ;;
    "arj")
  		EXTENSION="arj"
	  	COMANDO_COMPRIMIR="arj a"
		  COMANDO_DESCOMPRIMIR="arj x"
		  COMANDO_CONTENIDO="arj l"
		  ARGUMENTO_INCREMENTAL=""
		  ARGUMENTO_DESCOMPRIMIR="-C"
		  TODOS=""
		  BANDERA=2
    ;;
  esac
  return 0
}

# ------------------ Funcion de ayuda ---------------------

funcionAyuda(){
	"${ECHO}" >&2     "${SISTEMA}"" v""${VERSION}"
	"${ECHO}" >&2
	"${ECHO}" >&2     "Usos: backardi [comandos]"
	"${ECHO}" >&2     "      backardi --full"
	"${ECHO}" >&2     "      backardi --inc"
	"${ECHO}" >&2     "      backardi --res [archivo]"
	"${ECHO}" >&2     "      backardi --cont [archivo]"
	"${ECHO}" >&2     "      backardi --help"
	"${ECHO}" >&2
	"${ECHO}" >&2     "Comandos:"
	"${ECHO}" >&2     "Esto son alguno de los comandos o argumentos"
	"${ECHO}" >&2     "que se puede utilizar".
	"${ECHO}" >&2     " --full   -f               Realiza un Backup completo."
	"${ECHO}" >&2     " --inc    -i               Realiza un Backup incremental."
	"${ECHO}" >&2     " --res    -r  [archivo]    Restaura el archivo especificado."
	"${ECHO}" >&2     " --cont   -c  [archivo]    Muestra el contenido del archivo."
	"${ECHO}" >&2     " --help   -h               Para ver mas informacion."
	"${ECHO}" >&2
}

# ---------------- Aplicacion principal -------------------

case "$1" in 
	-f|--full)
		"${ECHO}" "--------------- Backup Full ---------------"
		controlCompresor
		controlExtension
		if [ "${BANDERA}" = 0 ] || [ "${BANDERA}" = 1 ] ; then {
			$COMANDO_COMPRIMIR $BACKUP_DEST_DIR/`$HOSTNAME`-$NOMBRE_FECHA-FULL.$EXTENSION $BACKUP_DIR$TODOS
		} else {
			$COMANDO_COMPRIMIR $BACKUP_DIR$TODOS | bzip2 > $BACKUP_DEST_DIR/`$HOSTNAME`-$NOMBRE_FECHA-FULL.$EXTENSION
		}
		fi
		"${ECHO}" "------------ Backup Finalizado ------------"
		;;
	-i|--inc)
		"${ECHO}" "--------------- Backup Inc ----------------"
		controlCompresor
		controlExtension
		if [ "${BANDERA}" = 0 ] ; then {
			find $BACKUP_DIR -mtime -1 -type f -print | $COMANDO_COMPRIMIR $BACKUP_DEST_DIR/`$HOSTNAME`-$NOMBRE_FECHA-INC.$EXTENSION $ARGUMENTO_INCREMENTAL
		}
		fi
		if [ "${BANDERA}" = 1 ] ; then {
			$COMANDO_COMPRIMIR $BACKUP_DEST_DIR/`$HOSTNAME`-$NOMBRE_FECHA-INC.$EXTENSION `find $BACKUP_DIR -mtime -1 -type f -print`
		}
		fi
		if [ "${BANDERA}" = 2 ] ; then {
			$COMANDO_COMPRIMIR `find $BACKUP_DIR$TODOS -mtime -1 -type f -print` | bzip2 > $BACKUP_DEST_DIR/`$HOSTNAME`-$NOMBRE_FECHA-INC.$EXTENSION
		} 
		fi
		"${ECHO}" "------------ Backup Finalizado ------------"
		;;
	-r|--res)
		"${ECHO}" "------------- Restaurar Backup ------------"
		controlCompresor
		controlExtension
		$COMANDO_DESCOMPRIMIR $2 $ARGUMENTO_DESCOMPRIMIR /
		"${ECHO}" "------- El sistema fue reestablecido ------"
		;;
	-c|--cont)
		controlCompresor
		controlExtension
		"${ECHO}" "---------- Contenido del archivo ----------"
		$COMANDO_CONTENIDO $2 | less
		"${ECHO}" "-------------------------------------------"
		;;
	-in|--info)
		"${ECHO}" "----------- Informacion General -----------"
		stat $2
		"${ECHO}" "-------------------------------------------"
		;;
	-V|--version)
		"${ECHO}" "${SISTEMA}"" v""${VERSION}"
		;;
	-h|--help)
		funcionAyuda
		;;
    *)
        $ECHO "Utilizar: -h o --help para mas informacion."
        ;;
    esac
exit 0
